File: leetcode/algorithms/762_prime_number_of_set_bits_in_binary_representation/main.js
/**
 * @param {number} n
 * @return {boolean}
 */
var isPrime = function (n) {
  if (n < 2) {
    return 0;
  }

  for (let i = 2; i < n; i++) {
    if (n % i == 0) {
      return 0;
    }
  }

  return 1;
};

/**
 * @param {number} left
 * @param {number} right
 * @return {number}
 */
var countPrimeSetBits = function (left, right) {
  let count = 0;

  for (let i = left; i <= right; i++) {
    let bits = 0;
    let n = i;

    while (n > 0) {
      bits += n % 2;
      n = Math.floor(n / 2);
    }

    if (isPrime(bits)) {
      count++;
    }
  }

  return count;
};


// Solution
var countPrimeSetBits = function (left, right) {
  let count = 0;
  while (left <= right) {
    count += (665772 >> left.toString(2).match(/1/g).length) & 1;
    left++;
  }
  return count;
};

File: leetcode/algorithms/762_prime_number_of_set_bits_in_binary_representation/main.dart
class PrimeNumberOfSetBitsInBinaryRepresentation {
  bool isPrime(int n) {
    if (n < 2) {
      return false;
    }

    for (int i = 2; i < n; i++) {
      if (n % i == 0) {
        return false;
      }
    }

    return true;
  }

  int countPrimeSetBits(int left, int right) {
    int count = 0;

    for (int i = left; i <= right; i++) {
      int bits = 0;
      int n = i;

      while (n > 0) {
        bits += n % 2;
        n ~/= 2;
      }

      if (isPrime(bits)) {
        count++;
      }
    }

    return count;
  }


  // Solution
  int _bitCount(int n) {
    int count = 0;

    while (n > 0) {
      n &= (n - 1);
      count++;
    }

    return count;
  }

  int solution(int left, int right) {
    int count = 0;

    for (int i = left; i <= right; i++) {
      count += (665772 >> _bitCount(i)) & 1;
    }

    return count;
  }
}
